concrete/authentication/concrete/controller.php

```diff
@@ -321,48 +329,52 @@ public function authenticate()
             throw new \Exception($ip_service->getErrorMessage());
         }
 
-        $user = new User($uName, $uPassword);
-        if (!is_object($user) || !($user instanceof User) || $user->isError()) {
-            switch ($user->getError()) {
-                case USER_SESSION_EXPIRED:
-                    throw new \Exception(t('Your session has expired. Please sign in again.'));
-                    break;
-                case USER_NON_VALIDATED:
-                    throw new \Exception(
-                        t(
-                            'This account has not yet been validated. Please check the email associated with this account and follow the link it contains.'));
-                    break;
-                case USER_INVALID:
-                    // Log failed auth
-                    $ip_service->logFailedLogin();
-                    if ($ip_service->failedLoginsThresholdReached()) {
-                        $ip_service->addToBlacklistForThresholdReached();
-                        throw new \Exception($ip_service->getErrorMessage());
-                    }
 
-                    if ($this->isPasswordReset()) {
-                        Session::set('uPasswordResetUserName', $this->post('uName'));
-                        $this->redirect('/login/', $this->getAuthenticationType()->getAuthenticationTypeHandle(), 'required_password_upgrade');
-                    }
+        $loginService = $this->app->make(LoginService::class);
 
-                    if (Config::get('concrete.user.registration.email_registration')) {
-                        throw new \Exception(t('Invalid email address or password.'));
-                    } else {
-                        throw new \Exception(t('Invalid username or password.'));
-                    }
-                    break;
-                case USER_INACTIVE:
-                    throw new \Exception(t(Config::get('concrete.user.deactivation.message')));
-                    break;
-            }
+
+        try {
+            $user = $loginService->login($uName, $uPassword);
+        } catch (UserException $e) {
+            $this->handleFailedLogin($loginService, $uName, $uPassword, $e);
+        }
+
+        if ($user->isError()) {
+            throw new \RuntimeException(t('Unknown login error occurred. Please try again.'));
```

Comment:
I'd use a `Concrete\Core\Error\UserMessageException` instance (or a subclass of it), otherwise concrete5 will display only a generic error unless the `concrete.debug.display_errors`configuration key is set to a non falsy value.

Reply:
:+1: